compiler/src/iree/compiler/Codegen/Common/GPU/GPUMaterializeEncoding.cpp

```diff
@@ -85,6 +86,49 @@ static SmallVector<int64_t> getTransposePermutation(IREE::GPU::MMAAttr mma,
   }
 }
 
+/// Reshapes the `src` to `targetType` type using tensor.collapse_shape +
+/// tensor.expand_shape. The tensor.expand_shape is returned. Note that it
+/// assumes the `outerRank` dimensions of `src` type and `targetType` are
+/// identical, and the number of elements in the inner dimensions are the same.
+static tensor::ExpandShapeOp reshapeForInnerDims(RewriterBase &rewriter,
+                                                 Value src,
+                                                 RankedTensorType targetType,
+                                                 int64_t outerRank) {
+  auto srcType = cast<RankedTensorType>(src.getType());
+  SmallVector<int64_t> intermediateShape(
+      srcType.getShape().take_front(outerRank));
+  for (auto [a, b] : llvm::zip_equal(
+           intermediateShape, targetType.getShape().take_front(outerRank))) {
+    assert(a == b &&
+           "expect outerRank dimensions matched between src and targetType");
+  }
+
+  ArrayRef<int64_t> innerShape = srcType.getShape().drop_front(outerRank);
+  assert(llvm::all_of(innerShape,
+                      [](int64_t v) { return !ShapedType::isDynamic(v); }) &&
```

Comment:
Where else would you expect this refactored function to be used? In this context the inner shape should be non-dynamic coming out from packOp?

Reply:
I was going to use it in Contraction op materialization then I found that it is hard to represent reasonable IRs at this moment. Then I wrote down the plan of the lowering which could use the method.

Also, I think you'll need it for unset_encoding part? We likely need the same trick to make HAL/Flow ops happy.

> In this context the inner shape should be non-dynamic coming out from packOp?

Yeah, I'm just being careful. I documented it in the comment and think that we can do the assertion for it. The assertion should be gone in release build. Do you think that the check is too expensive?

Reply:
Oh that makes sense!